File: snippets/go/develop/agentsmd/clients.go
package develop

import (
	"context"
	"time"

	restate "github.com/restatedev/sdk-go"
	restateingress "github.com/restatedev/sdk-go/ingress"
)

type Clients struct{}

func (Clients) ClientExamples() {
	// <start_here>
	restateClient := restateingress.NewClient("http://localhost:8080")

	// Request-response
	result, err := restateingress.Service[string, string](
		restateClient, "MyService", "MyHandler").
		Request(context.Background(), "Hi")
	if err != nil {
		// handle error
	}

	// One-way
	restateingress.ServiceSend[string](
		restateClient, "MyService", "MyHandler").
		Send(context.Background(), "Hi")

	// Delayed
	restateingress.ServiceSend[string](
		restateClient, "MyService", "MyHandler").
		Send(context.Background(), "Hi", restate.WithDelay(1*time.Hour))
	// <end_here>

	_ = result
}

File: snippets/java/src/main/java/develop/agentsmd/Clients.java
package develop.agentsmd;

import dev.restate.client.Client;
import dev.restate.common.InvocationOptions;
import develop.MyObject;
import develop.MyService;
import java.time.Duration;

public class Clients {

  public void clientExamples() {
    // <start_here>
    Client restateClient = Client.connect("http://localhost:8080");

    // Request-response
    String result = restateClient.service(MyService.class).myHandler("Hi");

    // One-way
    restateClient.serviceHandle(MyService.class).send(MyService::myHandler, "Hi");

    // Delayed
    restateClient
        .serviceHandle(MyService.class)
        .send(MyService::myHandler, "Hi", Duration.ofSeconds(1));

    // With idempotency key
    restateClient
        .virtualObjectHandle(MyObject.class, "Mary")
        .send(MyObject::myHandler, "Hi", InvocationOptions.idempotencyKey("abc"));
    // <end_here>
  }
}
